Add support for importing C++ namespaces

bricknerb opened the pull request:
This adds support for having different C++ `NameScope`s (and not just the main `Cpp` scope), and we keep a pointer to `clang::DeclContext` these scopes so we can look up C++ names in the right part of the AST.

C++ Interop Demo with a namespace:

```c++
// hello_world.h

namespace some_namespace {

void hello_world();

}  // namespace some_namespace
```

```c++
// hello_world.cpp

#include <cstdio>

namespace some_namespace {

void hello_world() { printf("Hello World!\n"); }

}  // namespace some_namespace
```

```carbon
// main.carbon

library "Main";

import Cpp library "hello_world.h";

fn Run() -> i32 {
  Cpp.some_namespace.hello_world();
  return 0;
}
```

```shell
$ clang -c hello_world.cpp
$ bazel-bin/toolchain/carbon compile main.carbon
$ bazel-bin/toolchain/carbon link hello_world.o main.o --output=demo
$ ./demo
Hello World!
```

Closes #5102

bricknerb committed:
Add support for importing C++ namespaces
This adds support for having different C++ `NameScope`s (and not just the main `Cpp` scope), and we keep a pointer to `clang::DeclContext` these scopes so we can look up C++ names in the right part of the AST.

C++ Interop Demo with a namespace:

```c++
// hello_world.h

namespace some_namespace {

void hello_world();

}  // namespace some_namespace
```

```c++
// hello_world.cpp

#include <cstdio>

namespace some_namespace {

void hello_world() { printf("Hello World!\n"); }

}  // namespace some_namespace
```

```
// main.carbon

library "Main";

import Cpp library "hello_world.h";

fn Run() -> i32 {
  Cpp.some_namespace.hello_world();
  return 0;
}
```

```shell
$ clang -c hello_world.cpp
$ bazel-bin/toolchain/carbon compile main.carbon
$ bazel-bin/toolchain/carbon link hello_world.o main.o --output=demo
$ ./demo
Hello World!
```

jonmeow commented on toolchain/sem_ir/name_scope.h:
```diff
@@ -278,8 +284,9 @@ class NameScope : public Printable<NameScope> {
-  bool is_cpp_scope_ = false;
+  // Set if this is the `Cpp` scope or a scope inside `Cpp`. Points to the
+  // matching Clang declaration context to look for names.
+  clang::DeclContext* cpp_decl_context_ = nullptr;
```
Why is this mutable instead of constant? Can you please detail in comments?

bricknerb replied:
`clang::Sema::LookupQualifiedName()` requires a mutable `DeclContext`.
Added a comment.

bricknerb replied:
FWIW, I've looked into why `clang::Sema::LookupQualifiedName()` requires a mutable `DeclContext` and it seems like there are a lot of dependencies on this being mutable, so I couldn't change that.

jonmeow replied:
Seems fine; I think the DeclID TODO should resolve this anyways.

jonmeow commented on toolchain/sem_ir/name_scope.h:
```diff
@@ -278,8 +284,9 @@ class NameScope : public Printable<NameScope> {
-  bool is_cpp_scope_ = false;
+  // Set if this is the `Cpp` scope or a scope inside `Cpp`. Points to the
+  // matching Clang declaration context to look for names.
+  clang::DeclContext* cpp_decl_context_ = nullptr;
```
Can you please add the same TODO here as for the NamedDecl pointer? That is, use of IDs instead of pointers should probably be considered.

Note, I think we should probably resolve this sooner rather than later. Noted in https://docs.google.com/document/d/1mJk92JUPzPNr4LSDUUvfhL9-9i0Dh707uKCCuw3LncU/edit?disco=AAABfiUQDE8

bricknerb replied:
Makes sense.
Done.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -250,17 +251,48 @@ static auto ImportFunctionDecl(Context& context, SemIR::LocId loc_id,
+                                SemIR::NameId name_id,
+                                clang::NamespaceDecl* clang_decl)
+    -> SemIR::InstId {
+  auto namespace_inst =
+      SemIR::Namespace{.type_id = GetSingletonType(
+                           context, SemIR::NamespaceType::SingletonInstId),
+                       .name_scope_id = SemIR::NameScopeId::None,
+                       .import_id = SemIR::InstId::None};
+  auto namespace_inst_and_loc = SemIR::LocIdAndInst(
+      Parse::AnyNamespaceId(loc_id.node_id()), namespace_inst);
+  auto namespace_id =
+      AddPlaceholderInstInNoBlock(context, namespace_inst_and_loc);
+  namespace_inst.name_scope_id =
+      context.name_scopes().Add(namespace_id, name_id, parent_scope_id);
+  ReplaceInstBeforeConstantUse(context, namespace_id, namespace_inst);
```
Can you please look into sharing code with `AddImportNamespace`, even if it means splitting a helper function?

Note this feels similar to when we discussed `AddImportNamespace` on #4873.

bricknerb replied:
Thanks, I'm reusing the code now.
Trying to split it didn't work well so I expanded its possible behavior to allow not associating the namespace with an import and having it not add the name to the scope.
Noted.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -250,17 +251,48 @@ static auto ImportFunctionDecl(Context& context, SemIR::LocId loc_id,
+                       .name_scope_id = SemIR::NameScopeId::None,
+                       .import_id = SemIR::InstId::None};
+  auto namespace_inst_and_loc = SemIR::LocIdAndInst(
+      Parse::AnyNamespaceId(loc_id.node_id()), namespace_inst);
```
Is the `loc_id` here the location of a namespace declaration, or the call site? I think associating a call site location with the generated namespace entity is incorrect.

Note, you may want to rebase to trunk: the code here is changing to better enforce node kinds, and this will no longer compile.

bricknerb replied:
Thanks! I've removed the location for now since I don't have a better location to provide for this namespace.
Added a TODO.

jonmeow commented on toolchain/sem_ir/name_scope.h:
```diff
@@ -212,10 +213,15 @@ class NameScope : public Printable<NameScope> {
   }
 
-  auto is_cpp_scope() const -> bool { return is_cpp_scope_; }
+  auto is_cpp_scope() const -> bool { return cpp_decl_context(); }
```
Can you please update formatter.cpp to note when a namespace is coming from C++?

bricknerb replied:
Done, but I was really not sure about the format for this.

jonmeow commented on toolchain/check/testdata/interop/cpp/no_prelude/namespace.carbon:
```diff
@@ -0,0 +1,336 @@
+// CHECK:STDOUT:
+// CHECK:STDOUT: imports {
+// CHECK:STDOUT:   %Cpp: <namespace> = namespace file.%Cpp.import_cpp, [concrete] {
+// CHECK:STDOUT:     .my_namespace = <unexpected>.inst20.loc7_6
```
We have some cases where `<unexpected>` will crop up in failing code, but it never should in passing code. This looks like a regression in this PR, can you please investigate?

bricknerb replied:
Resolved.

jonmeow commented on toolchain/check/testdata/interop/cpp/no_prelude/namespace.carbon:
```diff
@@ -0,0 +1,336 @@
+// CHECK:STDOUT:   %foo3: %foo3.type = struct_value () [concrete]
+// CHECK:STDOUT: }
+// CHECK:STDOUT:
+// CHECK:STDOUT: imports {
+// CHECK:STDOUT:   %Cpp: <namespace> = namespace file.%Cpp.import_cpp, [concrete] {
+// CHECK:STDOUT:     .my_namespace1 = <unexpected>.inst20.loc7_6
+// CHECK:STDOUT:   }
+// CHECK:STDOUT: }
```
Comparing with example Carbon code:

```
// --- importee.carbon

package A;

namespace N1;
namespace N1.N2;
fn N1.F1();
fn N1.N2.F2();

// --- importer.carbon

import A;

fn Test() {
  A.N1.F1();
  A.N1.N2.F2();
}
```

In the SemIR for `importer` for the above, I can find the namespaces:

```
// CHECK:STDOUT: imports {
// CHECK:STDOUT:   %A: <namespace> = namespace file.%A.import, [concrete] {
// CHECK:STDOUT:     .N1 = %N1
// CHECK:STDOUT:     import A//default
// CHECK:STDOUT:   }
// CHECK:STDOUT:   %A.N1: <namespace> = import_ref A//default, N1, loaded
// CHECK:STDOUT:   %N1: <namespace> = namespace %A.N1, [concrete] {
// CHECK:STDOUT:     .F1 = %A.F1
// CHECK:STDOUT:     .N2 = %N2
// CHECK:STDOUT:     import A//default
// CHECK:STDOUT:   }
// CHECK:STDOUT:   %A.F1: %F1.type = import_ref A//default, F1, loaded [concrete = constants.%F1]
// CHECK:STDOUT:   %A.N2: <namespace> = import_ref A//default, N2, loaded
// CHECK:STDOUT:   %N2: <namespace> = namespace %A.N2, [concrete] {
// CHECK:STDOUT:     .F2 = %A.F2
// CHECK:STDOUT:     import A//default
// CHECK:STDOUT:   }
// CHECK:STDOUT:   %A.F2: %F2.type = import_ref A//default, F2, loaded [concrete = constants.%F2]
// CHECK:STDOUT: }
```

Can you look into why the `imports` here don't include the namespaces that are being imported? I think it'd be helpful to ensure that the namespaces are assigned to the import block.

It might help to look at the `import_ref_ids` use in `AddImportNamespace`, it seems relevant here.

bricknerb replied:
Thanks!
I've added the namespace to `import_ref_ids` and it seems to be much more similar.
It also resolved the `<unexpected>` issues.

I think the main diff that is left now is from C++ imports not using `ImportIRInstId`.
Let me know if you think we should use them, though I believe it should be done in a separate PR.

jonmeow replied:
I agree locations are a more complex issue, and yes can be separated.

bricknerb replied:
ok, taking no further action here.

bricknerb committed:
Import `cpp_decl_context_` comment

bricknerb committed:
Merge branch 'main' into namespace

bricknerb committed:
Remove the import C++ namespace location

bricknerb committed:
Add C++ namespace ids to `import_ref_ids` so they're properly included in SemIR

bricknerb committed:
Mark C++ imported namespaces as such in the SemIR dump

bricknerb committed:
Update all tests SemIR dumps with the new  C++ mark for imported C++ namespaces

bricknerb committed:
Reuse `AddImportNamespace()` in `ImportNamespaceDecl()`
Expand the behavior of `AddImportNamespace()` to support not associating the namespace with an import and not add the namespace name to the scope.

jonmeow commented on toolchain/sem_ir/formatter.cpp:
```diff
@@ -1277,6 +1277,9 @@ class FormatterImpl {
 
   auto FormatArg(NameScopeId id) -> void {
+    if (sem_ir_->name_scopes().Get(id).is_cpp_scope()) {
+      out_ << "[C++], ";
```
For format, how about mirroring what `scope.import_ir_scopes` does inside `FormatNameScope`? That is, something like `import Cpp//...` (with the ellipsis, similar to preludes). That way, we'd have a consistent spot to look to see where names are coming from.

Note, the `[concrete]` adjacent to this is also specifically a constant value -- partly I'm also trying to avoid confusion around the meaning of the square brackets.

bricknerb replied:
SG. Done.

jonmeow commented on toolchain/check/import.h:
```diff
@@ -11,20 +11,41 @@
 
+enum class AddImportNamespaceNameBehavior : int8_t {
+  // Do not try to add the namespace name to the scope and don't check if it
+  // already exists. Used when we want to add the name outside this function.
```
When you say "this function", which function are you referring to? I don't see any uses outside of `AddImportNamespace`.

bricknerb replied:
Clarified. I hope.

jonmeow commented on toolchain/check/import.cpp:
```diff
@@ -89,53 +89,69 @@ static auto CopyNameFromImportIR(Context& context,
-          DiagnoseDuplicateName(context, name_id, import_id, prev_inst_id);
+  bool inserted = false;
+  SemIR::NameScope::EntryId entry_id = SemIR::NameScope::EntryId::None;
+  bool add_name = name_behavior != AddImportNamespaceNameBehavior::DontAdd;
```
It looks like what this is doing is putting the start and end in a conditional, so that only the middle runs. Had you considered extracting out the middle section of this function (lines 148-160 -- maybe also making `import_id` a parameter instead of the `function_ref`) to a helper, and having the `DontAdd` case call that helper directly? Was there something that made that not work well?

bricknerb replied:
Yes, this works now, thanks! Done.

jonmeow commented on toolchain/check/import.cpp:
```diff
@@ -89,53 +89,69 @@ static auto CopyNameFromImportIR(Context& context,
   return import_name_id;
 }
 
+static auto MakeImportedNamespaceLocIdAndInstId(
```
This overlaps with what I'm doing in #5120 (where I'm trying to remove the `UnsafeMake`), so there may be a collision. That said, the `Id` suffix doesn't match the return type, so maybe:

```suggestion
// Returns the LocIdandInst for the namespace.
static auto MakeImportedNamespaceLocIdAndInst(
```

bricknerb replied:
Thanks for the heads up!
Done.

jonmeow commented on toolchain/check/import.cpp:
```diff
@@ -89,53 +89,69 @@ static auto CopyNameFromImportIR(Context& context,
 }
 
+static auto MakeImportedNamespaceLocIdAndInstId(
+    Context& context, SemIR::InstId import_id, SemIR::Namespace& namespace_inst)
```
We typically pass small structures by value instead of by reference, for better performance. Also, I don't think this needs to be mutable.

```suggestion
    Context& context, SemIR::InstId import_id, SemIR::Namespace namespace_inst)
```

bricknerb replied:
Done.

jonmeow commented on toolchain/check/import.cpp:
```diff
@@ -147,19 +163,21 @@ auto AddImportNamespace(Context& context, SemIR::TypeId namespace_type_id,
   // namespace may invalidate the pointer above.
   parent_scope = &context.name_scopes().Get(parent_scope_id);
```
This is only used inside the new `if`, maybe it should be moved?

bricknerb replied:
Done.

bricknerb committed:
Change C++ import namespace formatting to be more similar to importing prelude

bricknerb committed:
Clarify comment on `AddImportNamespaceNameBehavior::DontAdd`

bricknerb committed:
Rename `MakeImportedNamespaceLocIdAndInstId()` to `MakeImportedNamespaceLocIdAndInst()` and add a comment

bricknerb committed:
Pass `Namespace` by value to `MakeImportedNamespaceLocIdAndInst()`

bricknerb committed:
In `AddImportNamespace()`, define `parent_scope` only for the `add_name` case

bricknerb committed:
Extract some of the logic of `AddImportNamespace` and rename it to `AddImportNamespaceAndName`. Call the extracted logic function `AddImportNamespace`.
Change its API back to be more similar to the original API and when adding a namespace use the new extracted function.

jonmeow commented on toolchain/check/import.h:
```diff
@@ -12,21 +12,42 @@
+// `diagnose_duplicate_namespace` is used when handling a cross-package import,
+// where an existing namespace is in the current package and the new namespace
+// is a different package.
+auto AddImportNamespaceAndName(
```
`Add...AndName` implies to me that this is adding a name that `AddImportNamespace` is not adding, but I don't think that's what the difference. Maybe `AddImportNamespaceToScope`, since the difference is whether it's added to parent_scope_id?

bricknerb replied:
Makes sense.
Done.

jonmeow commented on toolchain/check/import.cpp:
```diff
@@ -113,56 +154,38 @@ auto AddImportNamespace(Context& context, SemIR::TypeId namespace_type_id,
-  namespace_inst.name_scope_id =
-      context.name_scopes().Add(namespace_id, name_id, parent_scope_id);
-  ReplaceInstBeforeConstantUse(context, namespace_id, namespace_inst);
+  auto result = AddImportNamespaceAndNameResult{
```
```suggestion
  AddImportNamespaceAndNameResult result = {
```

[Style](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/cpp_style_guide.md#syntax-and-formatting): See the initialization section.

bricknerb replied:
Done.

jonmeow reviewed (approved):
Basically LG, just two small things I'd appreciate being addressed prior to merge.

bricknerb committed:
Rename `AddImportNamespaceAndName` to `AddImportNamespaceToScope`

bricknerb committed:
Style fix
